MIT CCX: Use CCX Keys

cewing opened the pull request:
This pull request implements the use of ccx-specific opaque keys (`CCXLocator` and `CCXBlockUsageLocator` throughout the codebase.  

### History

The request to implement custom opaque keys for ccx was first brought up in an architecture review of the PR referenced above.  It was felt that using such keys would allow ccx to better integrate with existing logging and analytics systems.  The outcome of this conversation was captured in jazkarta/edx-platform#43 

A separate PR (https://github.com/edx/edx-platform/pull/8259) has been opened to add the dependency upon the [custom key implementation](https://github.com/jazkarta/ccx-keys).  That PR is under review and should be merged prior to this.

During implementation of this code, @nedbat, @cpennington, @ormsbee, and @carsongee have all been pulled into discussions at various points.

The work has been overseen by @pdpinch and the team at MIT.

### Purpose

By creating a custom opaque-key implementation, the mechanism by which a user views CCX content (as opposed to the course from which the ccx was created) is simplified, and made to fall in line with existing mechanisms for retrieving course modules, assets and so on.  In addition, because there is a "locator" for a ccx that is distinct from that for the course, logging and analytics can register the use of CCX content separately from the use of the same materials for the original course.

### Impacts

This implementation adds a wrapper around the modulestore returned by `xmodule.modulestore.django.modulestore`.  When the CCX feature is enabled in settings, then wherever that method is called, a wrapped version of the modulestore will be returned.  This wrapper strips the ccx information out of locators and keys as queries are sent in to the modulestore and restores the information to the data that is returned.  Because of this, there is little in the system that is not impacted by this change.  For queries to the modulestore that use non-ccx keys or locations, this wrapper should be transparent, but it is present.

An additional implication of this work is that CCXs can not be created from legacy courses that use the pre-split deprecated form of course keys and locations. This restriction can be worked around by creating a re-run of such a course, which will convert it to the new form of storage and keys/locations.  It is therefore not considered a blocker.

In all other respects, the work should not have any direct impact on users.  CCX functionality continues to work as it did before this change.  There are no user-facing UI elements that have changed.

This work is integrated with previous CCX work, so it will be released initially only on edge, where CCX is already in use. 

### Manual Testing

In order to test these changes, you must begin with a course that uses the split-modulestore and has opaque-key course keys and locations (`course-v1:org+course+run...`).  Once that is in place, then you can follow these steps:

1.  Log in to the LMS an admin for the course, view the "instructor" tab and select "membership"
2.  In the course staff section, select the 'CCX Coach' role and add a user as a ccx coach.
3.  Log in as the CCX Coach user and visit the CCX Coach tab.
4.  Fill in a name for your ccx and submit to create it. You should be redirected to the Coach Dashboard view for the CCX.
5.  verify that when viewing the coach tab, you see a url that contains `ccx-v1:` and has `+ccx@n` in the identifier (`n` is the integer primary key of the CCX that was created). 
6.  verify that when viewing any other top level tab, you see only the course id (it should start with `course-v1:` and *not* contain any mention of ccx).
7.  From the coach tab schedule some course content, but not all the items from the original course (so you can easily tell which you are seeing).
8.  From the coach tab, enroll a user in your ccx.
9.  Log in to the LMS as the student just enrolled by the coach.  You should see both the original course and the CCX on your dashboard.  They should be visually distinct
10.  Click on the CCX to go to the course info page.  It should look identical to the original course info page.  The URL for the page should contain `ccx-v1:` and `+ccx@n`.  (this verifies that you are viewing the ccx).
11.  Click on the courseware tab, verify that you see *only* those course elements that were scheduled by the coach, and not any others.
12.  View and interact with the courseware.  Verify that everything works as expected.
13.  Check your progress report and verify that it operates as expected.
14.  Verify that the student cannot see the discussion tab (it should be hidden for any ccx).
14.  Log in as Coach and check the grade reports under `student admin` to verify that grades are being correctly calculated.  Update the grading policy and verify that it is applied to both the Coach grade views and the student progress view.

### Concerns

The modulestore wrapper is not well covered with tests.  We are not comfortable that we understand how best to approach the testing of this element.  We would appreciate any pointers in that regard.  

The management of CCX keys and locations versus Course keys and locations is a bit tricky and there are many possible edge cases we have yet to uncover.  There are a few places where manually unwrapping a ccx key is required to get course access information directly.  We would appreciate attention paid to this aspect.  If there are other places we have not yet found or thought of that might need special attention, please let us know.

It should be noted that at the time of this PR, there are 5 test failures and one error happening in my local development environment.  All of these are in the tests for the sysadmin dashboard and have to do with importing XML course content from git.  I believe the problem to be in my development environment, but if anyone (like @carsongee) has suggestions about fixing them should they also appear in jenkins, I'd appreciate it.

edx-webhook commented:
Thanks for the pull request, @cewing! I've created [OSPR-632](https://openedx.atlassian.net/browse/OSPR-632) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

bdero commented:
Since the middleware went away, could you remove the inclusion of it from `aws.py`? It was a snag along the way while deploying: https://github.com/jazkarta/edx-platform/blob/ccx-custom-ids/lms/envs/aws.py#L609

amir-qayyum-khan commented:
thanks @bdero , with this line https://github.com/jazkarta/edx-platform/blob/ccx-custom-ids/lms/envs/aws.py#L609 i got exception https://gist.github.com/amir-qayyum-khan/764551f52d7f0f8cc020. Fixed by removing 
```diff
MIDDLEWARE_CLASSES += ('ccx.overrides.CcxMiddleware',)
```

amir-qayyum-khan commented:
@cewing getting exception on studio side on new and old courses https://gist.github.com/amir-qayyum-khan/b3ea7a5ac47737fc537c

amir-qayyum-khan commented:
### These below test cases are working fine **verified**

- verify that when viewing the coach tab, you see a url that contains ccx-v1: and has +ccx@n in the identifier (n is the integer primary key of the CCX that was created).
- verify that when viewing any other top level tab, you see only the course id (it should start with course-v1: and not contain any mention of ccx).
- From the coach tab schedule some course content, but not all the items from the original course (so you can easily tell which you are seeing).
- From the coach tab, enroll a user in your ccx.
- Log in to the LMS as the student just enrolled by the coach. You should see both the original course and the CCX on your dashboard. They should be visually distinct
- Click on the CCX to go to the course info page. It should look identical to the original course info page. The URL for the page should contain ccx-v1: and +ccx@n. (this verifies that you are viewing the ccx).
- Click on the courseware tab, verify that you see only those course elements that were scheduled by the coach, and not any others.
- View and interact with the courseware. Verify that everything works as expected.
- Check your progress report and verify that it operates as expected.
- Verify that the student cannot see the discussion tab (it should be hidden for any ccx).
- Log in as Coach and check the grade reports under student admin to verify that grades are being correctly calculated. Update the grading policy and verify that it is applied to both the Coach grade views and the student progress view.

cpennington commented on common/lib/xmodule/xmodule/modulestore/django.py:
```diff
@@ -193,6 +193,10 @@ def modulestore():
         )
 
+        if settings.FEATURES.get('CUSTOM_COURSES_EDX'):
+            from ccx.modulestore import CCXModulestoreWrapper
```
Does this import need to be delayed to here? If `CCXModulestoreWrapper` is imported even if the feature is off, are there any adverse consequences?

cewing replied:
No, the code is without side effects.  I'll fix that.

cewing replied:
I restored this after changing it as suggested because it causes circular import errors when the import is at module level.  We must leave it like this, I guess.

cpennington replied:
I think circular import is because `ccx/utils.py` is importing `modulestore.django`.

cewing replied:
@cpennington: maybe, though I can't see how importing `ccx.modulestore` would cause `ccx.utils` to get imported too.  

At any rate, module scope import blows everything up, so I'm going to leave this where it is.  Do you want me to comment on it to ensure someone doesn't move it later?

cpennington replied:
Yeah, a comment would be great. (Figuring out the circular import would be awesome too, but isn't required for right now).

cpennington commented on lms/djangoapps/ccx/modulestore.py:
```diff
@@ -0,0 +1,312 @@
+        run=ccx_locator.run,
+        branch=ccx_locator.branch,
+        version_guid=ccx_locator.version_guid,
+    )
```
Isn't this just `ccx_locator.to_course_locator()`?

cewing replied:
Yes it is.  This was implemented first, and as it became apparent that it would be useful elsewhere I added that API to the key implementation.  I forgot to remove this, but will do so.

cpennington commented on lms/djangoapps/ccx/modulestore.py:
```diff
@@ -0,0 +1,312 @@
+def strip_ccx(val):
+    retval = val
+    ccx_id = None
+    if hasattr(retval, 'ccx'):
```
Do you need this check, since you have the `isinstance` checks?

cewing replied:
probably not.  Will review.

cpennington commented on lms/djangoapps/ccx/modulestore.py:
```diff
@@ -0,0 +1,312 @@
+    def __getattr__(self, name):
+        """pass missing attributes through to _modulestore
+        """
+        return getattr(self._modulestore, name)
```
Do you need matching `__setattr__` and `__delattr__` calls?

cewing replied:
Do I?  Are there places in the code where attributes of the modulestore are set or deleted?  If so, then I would certainly need that.  So far that hasn't appeared to be the case.

cpennington commented on lms/djangoapps/ccx/modulestore.py:
```diff
@@ -0,0 +1,312 @@
+        retval = self._modulestore._clean_locator_for_mapping(locator)
+        if ccx:
+            retval = restore_ccx_collection(retval, ccx)
+        return retval
```
Did the context manager approach to abstracting out this pattern not work?

cewing replied:
Haven't gotten to implementing it yet.  I believe it will work.

cpennington commented on lms/djangoapps/ccx/overrides.py:
```diff
@@ -25,43 +27,42 @@ def get(self, block, name, default):
+    if isinstance(course_id, CourseKey):
+        course_key = course_id
+    else:
+        course_key = CourseKey.from_string(course_id)
```
Any reason not to just document that `course_id` should always be a `CourseKey`, and then make sure that's true in the rest of the code?

cewing replied:
Doesn't that just push the check for this into some other location?  I chose this anticipating that there might be places where the method was called that I don't know about (maybe not now, but in the future) and coded defensively against that possibility.

cpennington replied:
Ideally, we'd push all parsing of `course_keys` to the edges of the system, so that the internals are always using `CourseKey` objects, rather than maybe objects, maybe strings.

cewing replied:
Okay, I accept that argument.  I'll make the change and fix the call sites where it's coming in as a string

cpennington commented on lms/djangoapps/ccx/overrides.py:
```diff
@@ -25,43 +27,42 @@ def get(self, block, name, default):
+    if isinstance(course_key, CCXLocator):
+        ccx_id = course_key.ccx
+        ccx = CustomCourseForEdX.objects.get(pk=ccx_id)
+    return ccx
```
Nit:

    if not isinstance(course_key, CCXLocator):
        return None

    return CustomCourseForEdX.objects.get(pk=course_key.ccx)

cewing replied:
If you prefer.  I lean toward having a single return statement.

cpennington replied:
I tend to prefer the other style, but that's definitely a personal preference, not an edx.org style requirement.

cpennington commented on lms/djangoapps/ccx/tests/test_ccx_modulestore.py:
```diff
@@ -0,0 +1,116 @@
+            [
+                ItemFactory.create(parent=chapter) for _ in xrange(2)
+            ] for chapter in chapters
+        ])
```
You can actually just write this as:

    [ItemFactory.create(parent=chapter) for _ in xrange(2) for chapter in chapters]

cewing replied:
good point.  too much keeping the old implementation here.  The flatten isn't needed.  Thanks.

cpennington commented on lms/djangoapps/ccx/tests/test_utils.py:
```diff
@@ -185,6 +187,7 @@ def test_auto_enroll_passed_correctly(self):
 class TestEnrollEmail(ModuleStoreTestCase):
     """tests for the enroll_email function from ccx.utils
     """
+    MODULESTORE = TEST_DATA_SPLIT_MODULESTORE
```
Do you depend on this being a `SplitMongoModulestore`? Or could this specification of a modulestore for this test just be left out?

cewing replied:
I don't know.  We put it here to ensure that the course created will have a split-mongo id, but I suppose that's not a hard requirement, since these tests do not try to create a CCXLocator at any point.  

On the other hand, being consistent in the modulestore used for tests might be helpful in documenting that ccx is not compatible with the legacy mongo implementation.  I'm certainly open to changing that opinion, though @cpennington

cewing replied:
In fact, @cpennington, the utils method under test *does* construct a CCXLocator in order to put the correct link into the email sent to enrolled students.  So this must be a split modulestore in order to prevent attempting to build a CCXLocator with deprecated Course IDs

cpennington replied:
Ok, makes sense.

cpennington commented on lms/djangoapps/ccx/views.py:
```diff
@@ -71,43 +71,60 @@ def wrapper(request, course_id):
 @cache_control(no_cache=True, no_store=True, must_revalidate=True)
 @coach_dashboard
-def dashboard(request, course):
+def dashboard(request, course, ccx=None):
```
Should `course` here just be a `CCXLocator`?

cewing replied:
Preserving other dashboard code, which uses the `coach_dashboard` decorator, the expected value for 'course' is an actual Course descriptor.  I could update all the other dashboard views to derive the course from the CCXLocator, but that would be repeating a lot of code many times instead of repeating a small bit of code once or twice.

cpennington replied:
It looks to me like all of the uses of `@coach_dashboard` expect a `course_key` and an optional `ccx`. However, it seems like they are all ccx specific, and thus don't make sense in the case where `ccx is None`. But maybe I just didn't read into their behavior close enough.

A larger question about `@coach_dashboard`: Would it be clearer to use a class-based view? I know we don't use them much in edx code right now, but they seem to be the future, django-wise, and I expect to be using more soon.

cewing replied:
@cpennington:  The two places where `ccx=None` makes sense are the `dashboard` view and the `create_ccx` view.  Since we are wrapping all the views in the same decorator, in order to handle ccx and course lookup and access control from a single location, it makes sense to allow the ccx to be `None` when needed.  

We could do a class-based view, I suppose, but I'm really not a big fan of them.  Especially for simple, direct views like this.  It's a lot of magic and indirection added for what I would consider to be minimal benefit.  I'm just old-school like that I guess.

cpennington commented on lms/djangoapps/ccx/views.py:
```diff
@@ -71,43 +71,60 @@ def wrapper(request, course_id):
+            course_key = ccx.course_id
+
         role = CourseCcxCoachRole(course_key)
         if not role.has_user(request.user):
```
This check should validate that the user is a coach for the ccx, right?

cewing replied:
Yes.  It should.

cewing replied:
@cpennington: I've added a check for this, but it happens *after* this code block here, so github is not picking it up as making this comment obsolete.

cpennington commented on lms/djangoapps/ccx/views.py:
```diff
@@ -142,18 +159,20 @@ def create_ccx(request, course):
 @cache_control(no_cache=True, no_store=True, must_revalidate=True)
 @coach_dashboard
-def save_ccx(request, course):
+def save_ccx(request, course, ccx=None):
```
Should this use a `ccx_locator`?

cewing replied:
No.  We are creating override values for the course using the ccx id.  We can take apart a ccx locator and get those values, but the combination of needing to do that here and in several other dashboard views makes using the `coach_dashboard` decorator a DRYer implementation, I think.

cewing replied:
@cpennington: any further thoughts on this?

cpennington replied:
Yeah, this is fine.

cpennington commented on lms/djangoapps/ccx/views.py:
```diff
@@ -367,17 +398,22 @@ def ccx_invite(request, course):
-    url = reverse('ccx_coach_dashboard', kwargs={'course_id': course.id})
+    url = reverse(
+        'ccx_coach_dashboard',
+        kwargs={'course_id': CCXLocator.from_course_locator(course.id, ccx.id)}
```
This call indicates that my question about the definition of `ccx_coach_dashboard` is a reasonable one, since this is passing the `CCXLocator`, not the `CourseLocator`.

cewing replied:
I think this is a six-of-one kind of question.  Yes, we need to construct a CCXLocator from values that were previously derived from a CCXLocator.  But in most of these views we also need to have the course descriptor, so is it cheaper, more DRY to do that operation once and this one many times, or that one many times and this once?  I think its a toss-up.

cewing replied:
@cpennington: any further thoughts on this?  I am inclined to leave it as-is.

cpennington commented on lms/djangoapps/courseware/access.py:
```diff
@@ -100,6 +104,8 @@ def has_access(user, action, obj, course_key=None):
 
     if isinstance(obj, CourseKey):
+        if isinstance(obj, CCXLocator):
+            obj = obj.to_course_locator()
```
Should there instead be a  `_has_access_ccx_key`, which for the moment simply strips off the CCX part of the locator and passes down to `_has_access_course_key`?

cewing replied:
I suppose we could do that.  Would that be preferable in your view?

cpennington replied:
Yeah, I think that would be more consistent.

cewing commented:
@bdero, @amir-qayyum-khan I have removed all references to the middleware, apologies for the trouble there.

@amir-qayyum-khan the failure in using a "legacy" course is expected.  This ccx key implementation is **not** compatible with legacy courses.  They must be updated to split mongo by doing a re-run.  

And as for the error in attempting to mutate an opaque key, I'll fix that.  Clearly a problem.

cewing commented:
@sarina, etal

I've rebased to the most recent master to help keep this mergable.  I am also addressing comments and will do a squash of all the interim commits as soon as things seem more stable.

cewing commented:
@cpennington, there are a number of your comments above that I have questions about.  When you have a chance, would you please review and get back to me?  I'm anxious to resolve these as soon as possible for my client.

sarina commented:
@cewing have you looked into the build failures? There's some fairly significant seeming failures.

cewing commented:
@sarina, The build is failing due to the five tests that I mentioned in the "concerns" section of my Pull Request cover letter.  I have, as yet, been unable to determine what is causing them and am hoping that perhaps @carsongee or someone familiar with that section of code might be able to help me.  That's why I asked for help in the cover letter.  

Is there anyone there who can look over the failing tests and help me to figure out what is going wrong?

sarina commented:
@cewing ah I was more concerned about [these failures](https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/9065/testReport/junit/nose.failure/Failure/runTest_2/) of which there are a few:
```
nose.failure.Failure.runTest (from nosetests)
Error Message

No module named ccx.modulestore
Stacktrace

Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 327, in run
    testMethod()
  File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/jenkins/edx-venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/jenkins/workspace/edx-platform-test-subset/common/lib/xmodule/xmodule/modulestore/perf_tests/test_asset_import_export.py", line 17, in <module>
    from xmodule.modulestore.xml_importer import import_course_from_xml
  File "/home/jenkins/workspace/edx-platform-test-subset/common/lib/xmodule/xmodule/modulestore/xml_importer.py", line 45, in <module>
    from xmodule.modulestore.django import ASSET_IGNORE_REGEX
  File "/home/jenkins/workspace/edx-platform-test-subset/common/lib/xmodule/xmodule/modulestore/django.py", line 24, in <module>
    from ccx.modulestore import CCXModulestoreWrapper
ImportError: No module named ccx.modulestore
```

bdero commented:
Hmm, so I just deployed this to the sandbox, and @pwilkins & I can't seem to view any studio courses (CCX or not, Split or not):

```python
Traceback (most recent call last):

  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 109, in get_response
    response = callback(request, *callback_args, **callback_kwargs)

  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 20, in _wrapped_view
    return view_func(request, *args, **kwargs)

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 257, in course_handler
    return course_index(request, CourseKey.from_string(course_key_string))

  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 20, in _wrapped_view
    return view_func(request, *args, **kwargs)

  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 91, in _wrapped_view
    response = view_func(request, *args, **kwargs)

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 487, in course_index
    course_structure = _course_outline_json(request, course_module)

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 326, in _course_outline_json
    include_children_predicate=lambda xblock: not xblock.category == 'vertical'

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 795, in create_xblock_info
    include_children_predicate=include_children_predicate,

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 1024, in _create_xblock_child_info
    ) for child in xblock.get_children()

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 795, in create_xblock_info
    include_children_predicate=include_children_predicate,

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 1024, in _create_xblock_child_info
    ) for child in xblock.get_children()

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 795, in create_xblock_info
    include_children_predicate=include_children_predicate,

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 1024, in _create_xblock_child_info
    ) for child in xblock.get_children()

  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 804, in create_xblock_info
    published = modulestore().has_published_version(xblock) if not is_library_block else None

  File "/edx/app/edxapp/edx-platform/lms/djangoapps/ccx/modulestore.py", line 244, in has_published_version
    xblock.scope_ids.usage_id

AttributeError: can't set attribute
```

Are you able to edit courses in studio locally @cewing ?

sarina commented:
Brandon - does your sandbox perhaps need migrations, or to be entirely
rebuilt? Sometimes I find that's the easiest way to resolve things.

On Wed, Jun 10, 2015 at 11:58 AM, Brandon DeRosier <notifications@github.com
> wrote:

> Hmm, so I just deployed this to the sandbox, and @pwilkins
> <https://github.com/pwilkins> & I can't seem to view any studio courses
> (CCX or not, Split or not):
>
> Traceback (most recent call last):
>
>   File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 109, in get_response
>     response = callback(request, *callback_args, **callback_kwargs)
>
>   File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 20, in _wrapped_view
>     return view_func(request, *args, **kwargs)
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 257, in course_handler
>     return course_index(request, CourseKey.from_string(course_key_string))
>
>   File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 20, in _wrapped_view
>     return view_func(request, *args, **kwargs)
>
>   File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 91, in _wrapped_view
>     response = view_func(request, *args, **kwargs)
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 487, in course_index
>     course_structure = _course_outline_json(request, course_module)
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/course.py", line 326, in _course_outline_json
>     include_children_predicate=lambda xblock: not xblock.category == 'vertical'
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 795, in create_xblock_info
>     include_children_predicate=include_children_predicate,
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 1024, in _create_xblock_child_info
>     ) for child in xblock.get_children()
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 795, in create_xblock_info
>     include_children_predicate=include_children_predicate,
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 1024, in _create_xblock_child_info
>     ) for child in xblock.get_children()
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 795, in create_xblock_info
>     include_children_predicate=include_children_predicate,
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 1024, in _create_xblock_child_info
>     ) for child in xblock.get_children()
>
>   File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py", line 804, in create_xblock_info
>     published = modulestore().has_published_version(xblock) if not is_library_block else None
>
>   File "/edx/app/edxapp/edx-platform/lms/djangoapps/ccx/modulestore.py", line 244, in has_published_version
>     xblock.scope_ids.usage_id
> AttributeError: can't set attribute
>
> Are you able to edit courses in studio locally @cewing
> <https://github.com/cewing> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8395#issuecomment-110810815>.
>

bdero commented:
We're good on the migrations front, and I'm hoping it's not the deployment since we run these roles in all envs. ;) I'm updating a devstack to this branch now to see if I can reproduce it locally.

cewing commented:
I will be hopping onto this right away and will try to verify this behavior locally.  I've been running all tests so I'm surprised by these outcomes, but I have not  tried running studio directly yet.  More in a short while

cewing commented:
@sarina: Looking at that stack trace, I wonder if we aren't running into a weird Django import order problem.  I had been importing the ccx modulestore wrapper in that `xmodule.modulestore.django` module right at the location where I needed it, and changed that to a module-level import at the suggestion of code review.  I will see if I can replicate those failures here, and see if changing the import back works to fix them.

cewing commented:
@bdero: I've checked here locally and I can edit existing courses both in the old mongo format and in the new split format.  I can create new courses and put material in them.  The new course I created is using the old mongo format.  I can do a re-run of that new course and the re-run is in the split format, as expected.  I can edit both of these without issue.  

Can we narrow down at all where the traceback you  are seeing is coming from so that I can duplicate it locally?  It would even help if you could drop a breakpoint into `/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py` at line 804 and tell me what type of object 'xblock' is.  Clearly I am trying to set the usage_id attribute, which is not settable.  But I'm not sure exactly how to fix that.  Some idea of what kind of object it is might help a great deal in solving that problem.

bdero commented:
@cewing I was unable to reproduce it locally as well. I'm deploying it to another sandbox to see how thing will fare out.

I'm able to drop into pdb on the sandbox. Here's what that xblock is before it blows up:
```python
> /edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/item.py(807)create_xblock_info()
-> published = modulestore().has_published_version(xblock) if not is_library_block else None
(Pdb) type(xblock)
<class 'xblock.internal.CourseDescriptorWithMixins'>
(Pdb) xblock
CourseDescriptorWithMixins(CombinedSystem(None, CachingDescriptorSystem(<xmodule.modulestore.mongo.draft.DraftModuleStore object at 0x94d2710>, u'fgh/fgh/fgh', [u'i4x://fgh/fgh/course/fgh'], <class 'xmodule.hidden_module.HiddenDescriptor'>, [])), <KvsFieldData MongoKeyValueStore({u'grading_policy': {u'GRADE_CUTOFFS': {u'Pass': 0.5}, u'GRADER': [{u'min_count': 12, u'weight': 0.15, u'type': u'Homework', u'drop_count': 2, u'short_label': u'HW'}, {u'min_count': 12, u'type': u'Lab', u'drop_count': 2, u'weight': 0.15}, {u'min_count': 1, u'weight': 0.3, u'type': u'Midterm Exam', u'drop_count': 0, u'short_label': u'Midterm'}, {u'min_count': 1, u'weight': 0.4, u'type': u'Final Exam', u'drop_count': 0, u'short_label': u'Final'}]}, u'wiki_slug': u'fgh.fgh.fgh'}, None, [], {u'tabs': [{u'type': u'courseware', u'name': u'Courseware'}, {u'type': u'course_info', u'name': u'Course Info'}, {u'type': u'textbooks', u'name': u'Textbooks'}, {u'type': u'discussion', u'name': u'Discussion'}, {u'type': u'wiki', u'name': u'Wiki'}, {u'type': u'progress', u'name': u'Progress'}], u'start': u'2030-01-01T00:00:00Z', u'display_name': u'fgh', u'language': u'en', u'discussion_topics': {u'General': {u'id': u'i4x-fgh-fgh-course-fgh'}}})<{}, {}>>, ScopeIds(user_id=None, block_type=u'course', def_id=BlockUsageLocator(CourseLocator(u'fgh', u'fgh', u'fgh', None, None), u'course', u'fgh'), usage_id=BlockUsageLocator(CourseLocator(u'fgh', u'fgh', u'fgh', None, None), u'course', u'fgh')))
```

Which seems innocent enough.. This is just after creating a brand new course. I'm going to poke around and see if I can find the deviation between the sandbox and my devstack.

cewing commented:
@bdero: thanks for that.  I have a suspicion about the problem.  

Can you try making this change in your sandbox where the failure is happening and see if it helps:

in `lms.djangoapps.ccx.modulestoreL242-249` change this:

```
    def has_published_version(self, xblock):
        xblock.scope_ids.usage_id, ccx = strip_ccx(
            xblock.scope_ids.usage_id
        )
        retval = self._modulestore.has_published_version(xblock)
        if ccx:
            retval = restore_ccx_collection(retval, ccx)
        return retval
```
 to this:
```
    def has_published_version(self, xblock):
        xblock.location, ccx = strip_ccx(xblock.location)
        retval = self._modulestore.has_published_version(xblock)
        if ccx:
            retval = restore_ccx_collection(retval, ccx)
        return retval
```

I think that xblock.location is a wrapper around xblock.scope_ids and the reason you can't set usage_id directly is that it and def_id (which is another attribute of scope_ids) need to stay synced.

cewing commented:
apologies for accidentally closing this PR.  That was not intentional.

@sarina, I've got the import errors solved.  Somehow, having the import of ccx.modulestore.CCXModulestoreWrapper at the module scope in that file causes errors.  I suspect Django import time magic.  Moving the import back to where it was before code review solves the problem.

cpennington commented:
@cewing, it looks to me like there's probably a circular import between
ccx.modulestore and modulestore.django (when you moved ccx.modulestore into
the top of the file).

-Cale

On Wed, Jun 10, 2015 at 4:33 PM, Cris Ewing <notifications@github.com>
wrote:

> apologies for accidentally closing this PR. That was not intentional.
>
> @sarina <https://github.com/sarina>, I've got the import errors solved.
> Somehow, having the import of ccx.modulestore.CCXModulestoreWrapper at the
> module scope in that file causes errors. I suspect Django import time
> magic. Moving the import back to where it was before code review solves the
> problem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8395#issuecomment-110903974>.
>

cpennington commented on lms/djangoapps/ccx/modulestore.py:
```diff
@@ -0,0 +1,282 @@
+
+    def _clean_locator_for_mapping(self, locator):
+        with remove_ccx(locator) as stripped:
+            locator, restore = stripped
```
I think you might even be able to do this as: `with remove_ccx(locator) as (locator, restore):`

cpennington commented on lms/djangoapps/ccx/modulestore.py:
```diff
@@ -0,0 +1,282 @@
+        with remove_ccx(locator) as stripped:
+            locator, restore = stripped
+            retval = self._modulestore._clean_locator_for_mapping(locator)
+            return restore(retval)
```
And then this could just be: `return restore(self._modulestore._clean_locator_for_mapping(locator))`

cpennington commented on lms/djangoapps/ccx/tests/test_ccx_modulestore.py:
```diff
@@ -0,0 +1,110 @@
+    """
+    For [[1, 2], [3, 4]] returns [1, 2, 3, 4].  Does not recurse.
+    """
+    return [x for sub in seq for x in sub]
```
dead code.

cpennington commented on lms/djangoapps/dashboard/sysadmin.py:
```diff
@@ -59,9 +60,13 @@ def __init__(self, **kwargs):
         self.is_using_mongo = True
-        if isinstance(self.def_ms, XMLModuleStore):
+        if isinstance(check_ms, XMLModuleStore):
             self.is_using_mongo = False
```
It would be nice to convert this over to use `get_modulestore_type`, since that's a cleaner interface for this check.

cewing replied:
@cpennington: not entirely sure how I missed that.  It's much more elegant than this code.  I will make that fix both here and in the tests below.

cpennington commented on lms/djangoapps/dashboard/tests/test_sysadmin.py:
```diff
@@ -315,8 +316,12 @@ def test_xml_course_add_delete(self):
+        if isinstance(check_ms, CCXModulestoreWrapper):
+            check_ms = check_ms._modulestore
+
+        self.assertIn('xml', str(check_ms.__class__))
```
As above, using `get_modulestore_type`.

cpennington commented:
Done with this pass.

cewing commented:
@sarina: this has been rebased again onto master as of 233aba74a3228833609b2202c29c677a749804d0 and all tests are passing.  There is a single quality violation for an import error that I believe to be a phantom.  I've addressed all the review comments raised thus far, and I have test coverage of at least 86% on all the code I've written.  

Anything else I can do to help get this merged?

sarina commented:
@cewing : @cpennington and the platform team are your reviewers, so you'll have to talk to them.

sarina commented:
I am concerned about the quality error, that usually doesn't happen. Perhaps Cale can provide some insight as to how to handle it.

cpennington commented:
@cewing the quality error is identifying (in some sense) that the dependency on CCX is going in the wrong direction (I hadn't noticed before, but `ccx` is in `lms/djangoapps`. Typically, we don't want things in `common/lib` to import from `lms` or `cms` subdirectories).

I think the best way around this would be to have the addition of the `CCXModulestore` be declared in settings (or perhaps via a list of modulestore wrappers, or something like that), so that the modulestore library doesn't have to depend on CCX directly. However, I don't want to block this particular PR on that.

Can you add a `pylint: disable=` directive to that import (to squelch the quality error), a JIRA ticket capturing the pending refactor, and a TODO that points to that ticket?

cewing commented:
@sarina, apologies for dragging you into this

@cpennington: that makes sense.  I can disable pylint and mark the code site with a TODO, but I do not have access to JIRA.  Perhaps @pdpinch can take care of that for us?

cewing commented:
@cpennington: cb431cc should remove that last quality failure.  I've made a JIRA account and will open an issue

cewing commented:
@cpennington: I have opened [OPEN-694](https://openedx.atlassian.net/browse/OPEN-694) to track this.

cewing commented:
@cpennington: We've got a green board!

cewing commented:
Thank you, @cpennington and everyone else!

cpennington commented on common/djangoapps/student/models.py:
```diff
@@ -1195,6 +1195,13 @@ def is_enrolled(cls, user, course_key):
+        from ccx_keys.locator import CCXLocator
+        if isinstance(course_key, CCXLocator):
+            course_key = course_key.to_course_locator()
+
```
@ormsbee just mentioned to me that he had talked to @pdpinch about this, and that they were hoping to actually use the serialized CCXCourseKeys in the enrollment object.

Also, I think you actually want this to be `CCXCourseKey`, not `CCXLocator` (since `CCXLocator` is a `UsageKey`, not a `CourseKey`)

cpennington replied:
Bah. Never mind that last bit. I misread the CCX code.

cewing replied:
@ormsbee, @pdpinch, @cpennington:

I believe that it is on the radar to completely remove the separate concept of a 'ccx membership' and replace it with an enrollment in a course identified by a `CCXLocator` (which is a [admittedly badly-named] CourseKey).  We have [an issue open](https://github.com/jazkarta/edx-platform/issues/78) in our github tracker for this.

In the interests of limiting the scope of the work to implement custom opaque keys for ccx, we decided to handle that as a separate task.  I agree completely that it should be done, and when it is, this customized code to look up access will be able to be deleted.

muzaffaryousaf commented on lms/djangoapps/ccx/modulestore.py:
```diff
@@ -0,0 +1,298 @@
+    def fill_in_run(self, course_key):
+        """See the docs for xmodule.modulestore.mixed.MixedModuleStore"""
+        with remove_ccx(course_key) as (course_key, restore):
+            return restore(self._modulestore.fill_in_run(course_key))
```
@cewing @cpennington I think we need to check for `self._modulestore` has `fill_in_run` attribute or not ? Some of the tests are failing at this line and the error is:
```
XMLModuleStore' object has no attribute 'fill_in_run'
```
Here is the line that causing failure: https://github.com/edx/edx-platform/pull/8082/files#diff-7cb89381438773a8f95fbfe764d06406R44

I'm already checking this but when it gets through the wrapper's fill_in_run call it fails.

Do you think we should add that check here too?

cewing replied:
It seems like the right way to handle this would be to pass calls to `hasattr` through to the modulestore contained by the wrapper.  Since that modulestore will be the eventual target of any call anyway, it'd be best to have it answer the question.

cpennington replied:
@cewing: yes, that makes sense.

@muzaffaryousaf what tests are you saying are failing? Were they unit tests that I missed? Tests on master? Manual testing?

cewing replied:
@muzaffaryousaf, @cpennington: although given that `hasattr` simply calls `getattr` and checks for an exception, I think that it won't do us any good.  

Alternatively, calling code could check the `get_modulestore_type` return value, but that seems hacky, as it implies that the calling code needs to know something about the implementation details of modulestores.  

@cpennington, what about a decorator for wrapper methods that checks the wrapped modulestore for the called method before executing, and raises `AttributeError` if not present?  Something like this (rough guess):

```python

def check_method(func):
    def inner(self, *args, **kwargs):
        try:
            getattr(self._modulestore, func.__name__)
        except AttributeError:
            raise
        else:
            return func(self, *args, **kwargs)
```

cewing replied:
@cpennington: They look to be tests from a new pull request.  They were not failing when we merged this PR

muzaffaryousaf replied:
@cpennington @cewing The tests are old but they are failing because we introduced a new bookmarks service in module render and that service is calling `fill_in_run` method & currently they are fail in case of `XMLModuleStore` objects.

muzaffaryousaf replied:
@cewing @cpennington  i was thinking something like this https://github.com/edx/edx-platform/blob/54f62a85f6164545151c32a063916cbeec94de46/common/lib/xmodule/xmodule/modulestore/mixed.py#L217 .

cpennington replied:
Yes, that handles the specific case of `fill_in_run`.

Looking at the possibilities around making the general case of `hasattr` just passing through to the contained modulestore, I think in this case, it's probably to just handle `fill_in_run` directly. (Or, perhaps, to just make `fill_in_run` a base method on modulestores that defaults to returning the supplied `course_key`).

cewing replied:
I am a bit concerned about API incompatibilities between the mixed modulestore and the underlying stores.  `fill_in_run` is one example. there may well be others.  It would probably be worth someone's time to work on addressing those so that cases like this can be avoided.

muzaffaryousaf replied:
@cewing @cpennington Any of you have any plan to make these changes ??
